flink-runtime/src/main/java/org/apache/flink/runtime/operators/JoinDriver.java

```diff
@@ -128,17 +130,22 @@ public void prepare() throws Exception{
 				ConfigConstants.DEFAULT_RUNTIME_HASH_JOIN_BLOOM_FILTERS);
 
 		// create and return joining iterator according to provided local strategy.
-		if (objectReuseEnabled) {
-			switch (ls) {
-				case INNER_MERGE:
-					this.joinIterator = new ReusingMergeInnerJoinIterator<>(in1, in2, 
+		if (reset) {
+			resetForIterativeTasks(in1, in2, serializer1, serializer2, comparator1, comparator2, pairComparatorFactory);
+			reset = false;
+		}
+		if (joinIterator == null) {
```

Comment:
Yes. I saw this pattern and how it was used. The existing ResettableDriver impls like
JoinWithSolutionSetFirstDriver, AbstractCachedBuildSideJoinDriver was actually doing things in 
`initialize()` was called only once for the first time and was leaving out `prepare()` as empty as it was called for every iteration. 
Now we may have to change even them if we need to unify all. I can make those changes. Let me check. Thanks @ggevay .